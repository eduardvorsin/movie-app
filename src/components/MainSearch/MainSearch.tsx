'use client';

import { GeneralProps } from '@/types/shared';
import { ChangeEventHandler, FormEventHandler, useReducer, useState } from 'react';
import { usePathname, useRouter } from 'next/navigation';
import { fallbackLng, Locales } from '@/i18n/settings';
import { MultiSearchData, ListsResponse } from '@/services/types';
import Button from '@/components/Button/Button';
import AutocompleteSearch, { AutocompleteOption, Props as AutocompleteSearchProps } from '@/components/AutocompleteSearch/AutocompleteSearch';
import { routes } from '@/constants';

export type MainSearchDictionary = {
	label: string,
	placeholder: string,
	openSearch: string,
	closeSearch: string,
	autocompleteSearch: AutocompleteSearchProps['dictionary'],
};

type Props = {
	id: string,
	onSearchClose: () => void,
	onSearchOpen: () => void,
	isSearchVisible: boolean,
	dictionary: MainSearchDictionary
} & GeneralProps;


type State = {
	status: 'loading' | 'successful' | 'error' | 'idle',
	items: AutocompleteOption[],
	error: string,
};

type Action = {
	type: 'set_loading'
} | {
	type: 'set_error',
	payload: string
} | {
	type: 'set_data',
	payload: AutocompleteOption[],
};

const createInitialState = (): State => {
	return {
		status: 'idle' as const,
		items: [],
		error: '',
	};
}

type Reducer = (state: State, action: Action) => State;

const reducer: Reducer = (state, action) => {
	if (action.type === 'set_loading') {
		return {
			status: 'loading' as const,
			error: '',
			items: [],
		};
	} else if (action.type === 'set_error') {
		return {
			status: 'error' as const,
			error: action.payload,
			items: [],
		}
	} else if (action.type === 'set_data') {
		return {
			status: 'successful' as const,
			error: '',
			items: action.payload,
		}
	}
	return state;
}

const transformToSearchResults = (data: MultiSearchData[]): AutocompleteOption[] => {
	if (!data) return [];

	const baseRoutes = {
		person: routes.persons,
		tv: routes.tv,
		movie: routes.movies,
	} as const;

	return data.slice(10).map((value) => {
		const { id, media_type } = value;

		return ({
			href: `${baseRoutes[media_type]}${id}`,
			label: value.media_type === 'movie' ? value.title : value.name,
			iconType: media_type,
		})
	});
};

export default function MainSearch({
	id,
	className,
	testId,
	isSearchVisible,
	onSearchClose,
	onSearchOpen,
	dictionary,
	...props
}: Props) {
	const [state, dispatch] = useReducer<Reducer, null>(reducer, null, createInitialState);
	const [value, setValue] = useState<string>('');
	const pathname = usePathname();
	const router = useRouter();

	const search = async (query: string, page: number, options?: {
		language: Locales
	}): Promise<void> => {
		const baseURL = '/api/multiSearchData';
		const searchParams = new URLSearchParams();
		searchParams.append('query', query);
		searchParams.append('page', page.toString());
		searchParams.append('lang', options?.language ?? fallbackLng);

		dispatch({ type: 'set_loading' });

		const res = await fetch(`${baseURL}?${searchParams}`);
		const data = (await res.json()) as ListsResponse<MultiSearchData> | null;

		if (data !== null && data?.results) {
			dispatch({
				type: 'set_data',
				payload: transformToSearchResults(data.results),
			});
		} else {
			dispatch({
				type: 'set_error',
				payload: 'Error receiving data from the server',
			});
		}
	};

	const clearSearchResults = (): void => setValue('');

	const onSearchChange: ChangeEventHandler<HTMLInputElement> = (e): void => {
		setValue(e.target.value);
		search(e.target.value, 1, { language: fallbackLng });
	};

	const onSearchSubmit: FormEventHandler<HTMLFormElement> = (e): void => {
		e.preventDefault();
		router.push(`${pathname.slice(0, 3)}/search?query=${value}`);
	};

	const classes = [
		className,
		'flex items-center justify-end flex-grow ',
		isSearchVisible ? 'h-full w-full lg:w-auto top-0 left-0 absolute px-4 py-3 lg:static lg:p-0 bg-neutral-100 dark:bg-dark-neutral-100 lg:bg-transparent lg:dark:bg-transparent' : '',

	].join(' ');

	return (
		<search
			role='search'
			className={classes}
			data-testid={testId}
			{...props}
		>
			{isSearchVisible ? (
				<AutocompleteSearch
					className='z-100 w-full'
					id={id}
					options={state.items}
					name='main-search'
					label={dictionary.label}
					value={value}
					placeholder={dictionary.placeholder}
					labelHidden
					isLoading={state.status === 'loading'}
					onChange={onSearchChange}
					onSubmit={onSearchSubmit}
					onClear={onSearchClose}
					onOptionLinkClick={clearSearchResults}
					dictionary={dictionary.autocompleteSearch}
				/>
			) : (
				<Button
					className='mr-1'
					appearance='secondary'
					iconButton
					onClick={onSearchOpen}
				>
					<svg
						className='p-2 w-[2.25rem] h-[2.25rem]'
						viewBox='0 0 32 32'
					>
						<use href={'/assets/icons/search.svg#search'} />
					</svg>
					{dictionary.openSearch}
				</Button>
			)}
		</search>
	);
}