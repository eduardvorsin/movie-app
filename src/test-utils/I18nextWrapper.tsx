import React, { ReactNode } from 'react';
import i18next from '@/i18n/client';
import { I18nextProvider } from 'react-i18next';

type Props = {
	children: ReactNode,
}

const I18nextWrapper = ({
	children,
}: Props) => {
	return (
		<I18nextProvider i18n={i18next}>
			{children}
		</I18nextProvider>
	);
};

export default I18nextWrapper;