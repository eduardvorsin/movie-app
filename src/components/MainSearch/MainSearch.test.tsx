import { render, screen } from "@testing-library/react";
import MainSearch from "./MainSearch";
import userEvent from "@testing-library/user-event";
import i18next from "@/i18n/client";
import I18nextWrapper from "@/test-utils/I18nextWrapper";

jest.mock('next/navigation', () => ({
	useParams: () => jest.fn(),
	useRouter: () => jest.fn(),
	usePathname: () => jest.fn(),
}));

const dictionary = {
	label: 'Search',
	placeholder: 'Search for movies, TV series, actors.',
	openSearch: 'Show the search field',
	closeSearch: 'Hide the search field',
	autocompleteSearch: {
		emptyStateTitle: 'We can\'t find anything for your request.',
		emptyStateText: 'Try entering a different value',
		search: {
			button: 'To find',
			clearButton: 'To clear',
		}
	},
};

describe('MainSearch tests', () => {
	it('is rendered correctly', () => {
		const closeSearchMockFn = jest.fn();
		const openSearchMockFn = jest.fn();

		render(
			<MainSearch
				id='main-search'
				isSearchVisible={false}
				onSearchClose={closeSearchMockFn}
				onSearchOpen={openSearchMockFn}
				dictionary={dictionary}
				testId='main-search'
			/>
		);

		expect(screen.getByRole<HTMLDivElement>('search')).toBeInTheDocument();
	});

	it('when you click on the "Clear" button, the mock function should be called', async () => {
		const closeSearchMockFn = jest.fn();
		const openSearchMockFn = jest.fn();
		const user = userEvent.setup();

		render(
			<MainSearch
				id='main-search'
				isSearchVisible={true}
				onSearchClose={closeSearchMockFn}
				onSearchOpen={openSearchMockFn}
				dictionary={dictionary}
				testId='main-search'
			/>
		);

		await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /clear/i }));

		expect(closeSearchMockFn).toHaveBeenCalledTimes(1);
	});

	it('when you click on the "Show the search field" button, the ioc function should be called', async () => {
		const closeSearchMockFn = jest.fn();
		const openSearchMockFn = jest.fn();
		const user = userEvent.setup();

		render(
			<MainSearch
				id='main-search'
				isSearchVisible={false}
				onSearchClose={closeSearchMockFn}
				onSearchOpen={openSearchMockFn}
				dictionary={dictionary}
				testId='main-search'
			/>
		);

		await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /show/i }));

		expect(openSearchMockFn).toHaveBeenCalledTimes(1);
	});

	it('is a basic snapshot', () => {
		const closeSearchMockFn = jest.fn();
		const openSearchMockFn = jest.fn();

		render(
			<MainSearch
				id='main-search'
				isSearchVisible={false}
				onSearchClose={closeSearchMockFn}
				onSearchOpen={openSearchMockFn}
				dictionary={dictionary}
				testId='main-search'
			/>
		);

		expect(screen.getByRole<HTMLDivElement>('search')).toMatchSnapshot();
	});

	it('is a snapshot with isSearchVisible set to true', () => {
		const closeSearchMockFn = jest.fn();
		const openSearchMockFn = jest.fn();

		render(
			<MainSearch
				id='main-search'
				isSearchVisible={true}
				onSearchClose={closeSearchMockFn}
				onSearchOpen={openSearchMockFn}
				dictionary={dictionary}
				testId='main-search'
			/>
		);

		expect(screen.getByRole<HTMLDivElement>('search')).toMatchSnapshot();
	});
});

describe('MainSearch integration tests', () => {
	it('localization into English works correctly', async () => {
		await i18next.changeLanguage('en');
		const closeSearchMockFn = jest.fn();
		const openSearchMockFn = jest.fn();
		const translation = new RegExp(i18next.t('mainSearch.openSearch'));
		const dictionary = {
			label: i18next.t('mainSearch.label'),
			placeholder: i18next.t('mainSearch.placeholder'),
			openSearch: i18next.t('mainSearch.openSearch'),
			closeSearch: i18next.t('mainSearch.closeSearch'),
			autocompleteSearch: {
				emptyStateTitle: i18next.t('autocompleteSearch.emptyStateTitle'),
				emptyStateText: i18next.t('autocompleteSearch.emptyStateText'),
				search: {
					button: i18next.t('search.button'),
					clearButton: i18next.t('search.clearButton'),
				},
			}
		};
		render(
			<MainSearch
				id='main-search'
				isSearchVisible={false}
				onSearchClose={closeSearchMockFn}
				onSearchOpen={openSearchMockFn}
				dictionary={dictionary}
				testId='main-search'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByRole<HTMLButtonElement>('button', { name: translation })).toBeInTheDocument();
	});

	it('localization into Russian works correctly', async () => {
		await i18next.changeLanguage('ru');
		const closeSearchMockFn = jest.fn();
		const openSearchMockFn = jest.fn();
		const translation = new RegExp(i18next.t('mainSearch.openSearch'));
		const dictionary = {
			label: i18next.t('mainSearch.label'),
			placeholder: i18next.t('mainSearch.placeholder'),
			openSearch: i18next.t('mainSearch.openSearch'),
			closeSearch: i18next.t('mainSearch.closeSearch'),
			autocompleteSearch: {
				emptyStateTitle: i18next.t('autocompleteSearch.emptyStateTitle'),
				emptyStateText: i18next.t('autocompleteSearch.emptyStateText'),
				search: {
					button: i18next.t('search.button'),
					clearButton: i18next.t('search.clearButton'),
				},
			}
		};
		render(
			<MainSearch
				id='main-search'
				isSearchVisible={false}
				onSearchClose={closeSearchMockFn}
				onSearchOpen={openSearchMockFn}
				dictionary={dictionary}
				testId='main-search'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByRole<HTMLButtonElement>('button', { name: translation })).toBeInTheDocument();
	});

	it('is a snapshot with English localization', async () => {
		await i18next.changeLanguage('en');
		const closeSearchMockFn = jest.fn();
		const openSearchMockFn = jest.fn();
		const dictionary = {
			label: i18next.t('mainSearch.label'),
			placeholder: i18next.t('mainSearch.placeholder'),
			openSearch: i18next.t('mainSearch.openSearch'),
			closeSearch: i18next.t('mainSearch.closeSearch'),
			autocompleteSearch: {
				emptyStateTitle: i18next.t('autocompleteSearch.emptyStateTitle'),
				emptyStateText: i18next.t('autocompleteSearch.emptyStateText'),
				search: {
					button: i18next.t('search.button'),
					clearButton: i18next.t('search.clearButton'),
				},
			}
		};
		render(
			<MainSearch
				id='main-search'
				isSearchVisible={false}
				onSearchClose={closeSearchMockFn}
				onSearchOpen={openSearchMockFn}
				dictionary={dictionary}
				testId='main-search'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByRole<HTMLDivElement>('search')).toMatchSnapshot();
	});

	it('is a snapshot with Russian localization', async () => {
		await i18next.changeLanguage('ru');
		const closeSearchMockFn = jest.fn();
		const openSearchMockFn = jest.fn();
		const dictionary = {
			label: i18next.t('mainSearch.label'),
			placeholder: i18next.t('mainSearch.placeholder'),
			openSearch: i18next.t('mainSearch.openSearch'),
			closeSearch: i18next.t('mainSearch.closeSearch'),
			autocompleteSearch: {
				emptyStateTitle: i18next.t('autocompleteSearch.emptyStateTitle'),
				emptyStateText: i18next.t('autocompleteSearch.emptyStateText'),
				search: {
					button: i18next.t('search.button'),
					clearButton: i18next.t('search.clearButton'),
				},
			}
		};
		render(
			<MainSearch
				id='main-search'
				isSearchVisible={false}
				onSearchClose={closeSearchMockFn}
				onSearchOpen={openSearchMockFn}
				dictionary={dictionary}
				testId='main-search'
			/>,
			{ wrapper: I18nextWrapper }
		);

		expect(screen.getByRole<HTMLDivElement>('search')).toMatchSnapshot();
	});
});