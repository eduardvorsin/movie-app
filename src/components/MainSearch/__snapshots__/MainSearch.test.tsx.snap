// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainSearch integration tests is a snapshot with English localization 1`] = `
<search
  class=" flex items-center justify-end flex-grow  "
  data-testid="main-search"
  role="search"
>
  <button
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   mr-1"
  >
    <span
      class="inline-flex items-center"
    >
      <svg
        class="p-2 w-[2.25rem] h-[2.25rem]"
        viewBox="0 0 32 32"
      >
        <use
          href="/assets/icons/search.svg#search"
        />
      </svg>
      Show the search field
    </span>
  </button>
</search>
`;

exports[`MainSearch integration tests is a snapshot with Russian localization 1`] = `
<search
  class=" flex items-center justify-end flex-grow  "
  data-testid="main-search"
  role="search"
>
  <button
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   mr-1"
  >
    <span
      class="inline-flex items-center"
    >
      <svg
        class="p-2 w-[2.25rem] h-[2.25rem]"
        viewBox="0 0 32 32"
      >
        <use
          href="/assets/icons/search.svg#search"
        />
      </svg>
      Показать поле поиска
    </span>
  </button>
</search>
`;

exports[`MainSearch tests is a basic snapshot 1`] = `
<search
  class=" flex items-center justify-end flex-grow  "
  data-testid="main-search"
  role="search"
>
  <button
    class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   mr-1"
  >
    <span
      class="inline-flex items-center"
    >
      <svg
        class="p-2 w-[2.25rem] h-[2.25rem]"
        viewBox="0 0 32 32"
      >
        <use
          href="/assets/icons/search.svg#search"
        />
      </svg>
      Show the search field
    </span>
  </button>
</search>
`;

exports[`MainSearch tests is a snapshot with isSearchVisible set to true 1`] = `
<search
  class=" flex items-center justify-end flex-grow  h-full w-full lg:w-auto top-0 left-0 absolute px-4 py-3 lg:static lg:p-0 bg-neutral-100 dark:bg-dark-neutral-100 lg:bg-transparent lg:dark:bg-transparent"
  data-testid="main-search"
  role="search"
>
  <div
    class="relative z-100 w-full"
  >
    <div>
      <form
        autocomplete="off"
      >
        <label>
          <span
            class="block text-100 mb-1 text-neutral-1000 dark:text-dark-neutral-1000 transition-colors duration-150 sr-only"
          >
            Search
          </span>
          <div
            class="relative"
          >
            <input
              class="outline-none w-full h-[2.25rem] py-2 bg-neutral-0 dark:bg-dark-neutral-200 border-neutral-300 dark:border-dark-neutral-350 text-neutral-1000 dark:text-dark-neutral-900 rounded rounded-[0.1875rem] border-2 text-100 enabled:hover:bg-neutral-100 enabled:hover:border-neutral-400 dark:enabled:hover:bg-dark-neutral-250 dark:enabled:hover:border-dark-neutral-350 focus:border-blue-500 dark:focus:border-blue-300 transition-colors duration-150   border-neutral-300 dark:border-dark-neutral-350 px-8"
              id="main-search"
              inputmode="search"
              name="main-search"
              placeholder="Search for movies, TV series, actors."
              type="text"
              value=""
            />
            <button
              class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-neutral-600 enabled:hover:text-neutral-700 enabled:active:text-neutral-800 dark:text-neutral-300 dark:enabled:hover:text-neutral-200 dark:enabled:active:text-neutral-100   [&]:text-neutral-400 enabled:hover:[&]:text-neutral-300 enabled:active:[&]:text-neutral-200 dark:[&]:text-dark-neutral-600 dark:enabled:hover:text-dark-neutral-700 dark:enabled:active:text-dark-neutral-800 absolute left-[0.375rem] top-1/2 -translate-y-1/2"
              type="button"
            >
              <span
                class="inline-flex items-center"
              >
                <svg
                  class="w-6 h-6 fill-current"
                  viewBox="0 0 32 32"
                >
                  <use
                    href="/assets/icons/cancel.svg#cancel"
                  />
                </svg>
                To clear
              </span>
            </button>
            <button
              class="items-center font-medium rounded-1 text-center cursor-pointer min-w[2.25rem] leading-2 inline-flex transition-colors duration-150 text-0  text-dark-neutral-0 enabled:hover:text-dark-neutral-400 enabled:active:text-dark-neutral-600 dark:text-neutral-0 dark:enabled:hover:text-neutral-400 dark:enabled:active:text-neutral-600   absolute top-1/2 right-[0.375rem] -translate-y-1/2"
            >
              <span
                class="inline-flex items-center"
              >
                <svg
                  class="w-6 h-6 p-0.5 fill-current"
                  viewBox="0 0 32 32"
                >
                  <use
                    href="/assets/icons/loupe.svg#loupe"
                  />
                </svg>
                To find
              </span>
            </button>
          </div>
        </label>
      </form>
    </div>
  </div>
</search>
`;
